### PR TITLE
layout component to wrap routes

### PR DESCRIPTION
**Pull request by MrLoh**
As described in #172 it would be useful to have a way to pass a component that wraps around the Routes, to implement custom route transitions and enable setting things like headers that appear on all pages. 

See example usage: 
https://github.com/MrLoh/afterjs-apollo-rnw-styled-pose-example/tree/layout

### CODE REVIEW

**Comment by MrLoh** [comment]
Ideally Layout would be of type `AsyncRouteComponentType` and also allow fetching data in a `getInitialProps` method. I'm not sure when the method would be called though.

**Comment by jaredpalmer** [comment]
Can you update from master?

**Comment by MrLoh** [comment]
@jaredpalmer I fixed the type errors, it's working fine for me. What are your thoughts on the provided Interface?

**Comment by odincov** [comment]
@jaredpalmer any plans to release this?

**Comment by DPr00f** [comment]
I guess it's only missing an example or a Readme file update. When should we expect this @jaredpalmer ?

**Comment by MrLoh** [comment]
I already have an example, would just have to remove a bunch of other stuff from it.
https://github.com/MrLoh/afterjs-apollo-rnw-styled-pose-example/tree/layout

Replicating this pose example would be nice I thing:
https://codesandbox.io/s/wq324qk687?from-embed

File: src/types.ts [hunk 1]
```diff
@@ -4,6 +4,13 @@ import { Request, Response } from 'express';
 import { IncomingMessage, ServerResponse } from 'http';
 import { History, Location } from 'history';
 
+export interface LayoutProps { 
+  location: Location; 
+  children: React.ReactNode 
+}
+
+export type LayoutComponent = React.ComponentClass<LayoutProps>
```
Comment (dagda1) [inline src/types.ts]: Could we make this `ComponentType<LayoutProps>` so the layout can be stateless?

Comment (MrLoh) [reply src/types.ts]: Sure, but I need a proper review from @jaredpalmer before I'm gonna touch the code again.

**Comment by nealoke** [comment]
The lack of being able to nest routes as you normally would in `react-router` and thus allowing `Layout` components as wrappers is a definite deal breaker to using `after.js`. Also being able to only define 1 `wrapper` component seems very limiting as mentioned at #47 which was just closed without any explanation.

Why not use a more `react-router-config` approach? 

## `Routes.js`
```
import Layout from "./Layout";
import Page from "./Page";

export default [
	{
		component: Layout,
		routes: [
			{
				component: Page,
				path: "/",
				exact: true
			}
		]
	}
];

```
## `Layout.js`
```
import { renderRoutes } from "react-router-config";

const Layout = ({ route }) => (
	<div>
		<div>{renderRoutes(route.routes)}</div>
	</div>
);
```

## `Page.js`
```
const Page = () => (
	<div>
		This will be wrapped by the Layout component
	</div>
);
```

**Comment by MrLoh** [comment]
@nealoke that sounds like an interesting idea, why don't you make a PR based on this one, that show how this could work.

**Comment by MrLoh** [comment]
But you can also already achieve the same thing now, by putting a `Route` component inside the Layout component to render different layouts for different routes.

**Comment by dagda1** [comment]
I could really do with this, i have a workaround I am currently using

**Comment by MrLoh** [comment]
It would be great to see this move forward. There hasn’t been any relevant comment as to why this wouldn’t be a good API yet. @jaredpalmer how can we get this unstuck?

**Comment by MrLoh** [comment]
If no layout component is specified it simply passes through the children (see After.tsx:105) so it doesn’t really need a default. I don’t know why this wouldn’t be isomorphic, I’ve been using it in production for a while now and it renders on both server and client side. The any types came from some complains the linter or typescript gave me, they could probably be removed.

**Comment by jaredpalmer** [comment]
Hrmmm. The more I think about it the more I think the better approach is react router config as it has baked in support for this. We just need to alter the the route walker slightly

**Comment by nealoke** [comment]
> But you can also already achieve the same thing now, by putting a `Route` component inside the Layout component to render different layouts for different routes.

@MrLoh I tried with below code but when navigating to `/page` it renders nothing. Am I doing this wrong?

```
import React, { PureComponent } from "react";
import { Route } from "react-router-dom";

class FirstPage extends PureComponent {
	render() {
		return <div>First page</div>;
	}
}

class SecondPage extends PureComponent {
	render() {
		return <div>Second page</div>;
	}
}

export default class Layout extends PureComponent {
	render() {
		return (
			<div>
				<Route exact path="/" component={FirstPage} />
				<Route path="/page" component={SecondPage} />
			</div>
		);
	}
}
```

**Comment by MrLoh** [comment]
@nealoke The code looks fine, but it is also way too incomplete to say why nothing is displaying and has no code that relates to this PR, please open a different issue (and provide a code sandbox example) to discuss this, we should keep this PR to the discussion of the new feature architecture.

**Comment by MrLoh** [comment]
@jaredpalmer I'm not sure how you envision the [react router config](https://github.com/ReactTraining/react-router/tree/master/packages/react-router-config) approach being used here. Do you mean to switch after to use `react-router-config` internally, to allow nested route config? I guess that makes sense, as it relies on an existing standard. However, this seems like a different issue, since I don't see anything about wrapping layout components in react-router-config. Or are you envisioning to completely change the After api, to expose `renderRoutes` in some way to the consumer of after? I'd be happy to help with a different implementation approach, but I'm not sure I understand the direction yey.

**Comment by MrLoh** [comment]
@jaredpalmer would be great if we can get this unstuck, I'm happy to work more on it. But I need input in the discussion

**Comment by oomathias** [comment]
Hi @MrLoh @jaredpalmer!

Here is what I did to support an App Layout (and tried to support customs route components, but it requires some API changes). I think `after.js` could be simplified and fully support `react-router` with a few changes.

```js
// App.js (this is a new file)
import React from 'react'
import { After } from '@jaredpalmer/after'
import { Route } from 'react-router-dom'

const App = ({ data, routes }) => (
  // Custom Layout
  <div>
    // Isomorphic AfterParty
    <After data={data} routes={routes} />
    // Custom route component example
    {routes.map((route, index) => (
      <Route
        key={`${index}`}
        path={route.path}
        exact={route.exact}
        component={route.footer}
      />
    ))}
  </div>
)

export default App
```

```js
// client.js
[...]
ensureReady(routes).then(data =>
  hydrate(
    <BrowserRouter>
      // We pass data and routes
      <App data={data} routes={routes} />
    </BrowserRouter>,
    document.getElementById('root')
  )
)
[...]
```

```js
// server.js
// Nothing special
```

```js
// Custom Document.js
[...]
export default class Document extends React.Component {
  static async getInitialProps({ assets, data, renderPage }) {
    // We ignore After sent by render, I don't see why it's there, did I miss something like caching?
    // see: https://github.com/jaredpalmer/after.js/blob/master/src/render.tsx#L44
    // {fn(After)({ routes, data })} could be just {fn({ routes, data })} and we shall provide a default App.js
    const page = await renderPage(_ => ({ data, routes }) =>
      <App data={data} routes={routes} />
    return { assets, data, ...page }
  }
  [...]
}
```
```js
// routes.js

import React from 'react'
import { asyncComponent } from '@jaredpalmer/after'

import NotFound from './pages/NotFound'

export default [
  {
    path: '/',
    exact: true,
    component: asyncComponent({
      loader: () => import('./pages/Home'),
      Placeholder: () => <div>...LOADING...</div>,
    }),
  },
  {
    path: '/about',
    exact: true,
    component: asyncComponent({
      loader: () => import('./pages/About'),
      Placeholder: () => <div>...LOADING...</div>,
    }),
    // Custom route component example (requires API changes to actually call getInitialProps)
    footer: asyncComponent({
      loader: () => import('./components/Footer'),
      Placeholder: () => <div>...LOADING...</div>,
    }),
  },
  {
    path: '**',
    component: NotFound,
  },
]
```

As is App.js works (custom layout), the custom route component doesn't.

To summarize by implementing a default `App.js` and delete the imported `After` from `render.js`, we could be able to fully support `react-router`. Again, maybe I missed something here: https://github.com/jaredpalmer/after.js/blob/3e4f2f4151ee656eff87faf9470cb77b86776167/src/render.tsx#L44

Stop here, if you don't want to know about custom route component as is doesn't really apply in this PR, I'm just thinking ahead.

Custom route component could be easily achievable with support of `getInitialProps`!
It will require few changes in the API.

```js
// After.defaultProps = { component: 'component' }
<After data={data} route={data} component='footer' />
```
`loadInitialProps`must also be modified to accept a component variable. (access to route[component] instead of route.component)

The only problem I see is loadInitialProps server side which must load all custom components route. It could loop over `Object.keys(route)` and look for a Component or AsyncComponent and construct a data object which could look like this:

AfterData data object
```js
{
  component: {...},
  footer: {...}
}
```

App.js could look like this in the end (instead of the one provided at the beginning)
```js
// App.js
const App = ({ data, routes }) => (
  // Custom Layout
  <div>
    // Isomorphic AfterParty
    <After data={data.component} routes={routes} />
    // Custom route component example
    <After data={data.footer} routes={routes} component='footer'/>
  </div>
```

Both route.component and route.footer component could have their own `getInitialProps`.
There could be side effect inside After `componentWillReceiveProps`, like multiple calls to https://github.com/jaredpalmer/after.js/blob/3e4f2f4151ee656eff87faf9470cb77b86776167/src/After.tsx#L38, maybe some logic could be extracted in another component that could be included inside App.

Of course, you can include whatever you want in `App.js` like `ThemeProvider`, different `Layout` for some routes (route.layout = something), support for a root component in routes.js ...

Let me know what you think, I could make start a PR.

**Comment by shaodahong** [comment]
yeh, this can resolve children route `getInitialProps` method don't execute, In Project

```js
class Layout extends React.Component {
  static async getInitialProps({ req, res, match }) {
    const item = await CallMyApi(`/v1/item${match.params.id}`);
    return { item };
  }

  render() {
    return (
      <div>
        <Nav />
        <Route
          path="/home"
          exact
          component={Home}
        />
        <Route
          path="/about"
          exact
          component={About}
        />
        <Footer />
      </div>
    );
  }
}
```

`Nav` and `Footer` is Layout common component, `Home` or `About` is main content, when switch route, shouldbe change `Home` or `About`, but now `Home` `getInitialProps` method don't execute

or provide a scheme for children route `getInitialProps` execution

**Comment by MrLoh** [comment]
Not stale, needs input from @jaredpalmer on design, I have no capacity to work on iterating the architecture here. I can implement a different architecture if there is a concrete proposal that will be accepted, but otherwise, this is my proposal. Take it or leave it. Please close if there are no plans on implementing this, but this decision should not be made by a bot.

**Comment by otaviosoares** [comment]
It would be awesome to see this move forward. I'm still getting familiar with after.js but I'm willing to contribute when there is an agreement on which path to follow.

**Comment by jaredpalmer** [comment]
I think this PR is on the right track. My hot take is that we should leverage `react-router-config` so that you can write a route tree like this:

> Note that `<App>` is the layout component we've been describing.

```tsx
import React from 'react';

import { asyncComponent } from '../asyncComponent';
import NonDymamicImport from './components/NonDynamicExport';
import { NoGetInitialProps } from './components/NoGetInitialProps';
import { App } from './components/App';

const Placeholder = () => <div>...LOADING...</div>;

export default [
  {
    path: '/',
    component: App,
    routes: [
      {
        path: '/',
        exact: true,
        component: asyncComponent({
          loader: () => import('./components/Home'),
          Placeholder,
        }),
      },
      {
        path: '/async-get-initial-props',
        exact: true,
        component: asyncComponent({
          loader: () => import('./components/AsyncGetInitialProps'),
          Placeholder,
        }),
      },
      {
        path: '/non-dynamic-import',
        exact: true,
        component: NonDymamicImport,
      },
      {
        path: '/non-default-export',
        exact: true,
        component: asyncComponent({
          loader: () =>
            import('./components/NoNDefaultExport').then(
              module => module.NonDefaultExport
            ),
          Placeholder,
        }),
      },
      {
        path: '/no-get-initial-props',
        exact: true,
        component: NoGetInitialProps,
      },
      {
        path: '/route',
        component: asyncComponent({
          loader: () => import('./components/RootRoute'),
          Placeholder,
        }),
        routes: [
          {
            path: '/route/nested',
            component: asyncComponent({
              loader: () => import('./components/NestedRoute'),
              Placeholder,
            }),
          },
        ],
      },
    ],
  },
];
```

`<App/>` is a layout component that must render the after party (i.e. where the data state is kept). This is the place for creating isomorphic, non-data context providers (e.g. react-helmet, react-gateway, etc.)

```tsx
import * as React from 'react';
import { renderRoutes, RouteConfigComponentProps } from 'react-router-config';
import { Helmet } from 'react-helmet'

export interface AppProps extends RouteConfigComponentProps<{}> {}

export interface AppState {}

export class App extends React.Component<AppProps, AppState> {
  render() {
    const { route, ...rest } = this.props;
    // route.route moves us into the children of the layout component for data fetching purposes
    return <>
              <Helmet titleTemplate="%s -Jared Palmer">           
                <meta property="og:site_name" content="Jared Palmer" />
                <meta name="og:url" content="..." />
                <meta name="og:image" content="/me.jpg" />
                <meta name="twitter:card" content="summary" />
                <meta name="twitter:url" content={url} />
                <meta name="twitter:site" content="@jaredpalmer" />
                <meta name="twitter:image" content="/me.jpg" />              
              </Helmet>
              <After routes={route.route} {...rest} />             
         </>;
  }
}
```

We may want to extend some helper callbacks for stuff like route changes out of the `<After>` component. This could be useful for analytics.

With this, we should simplify the code in render. We can also make the constraint that the layout component (i.e. `<App>` cannot be lazily loaded (since it is always rendered) anyways.

**Comment by jaredpalmer** [comment]
This PR is on the right track, but we should alter it so that you can do the route tree like I described above instead of.

**Comment by oomathias** [comment]
@jaredpalmer Is there a specific reason why you pass `After` here:
https://github.com/jaredpalmer/after.js/blob/ac1bb58771700e57fc1cf7ce94d43f230b33aca7/src/render.tsx#L44

If we create an isomorphic `<App/>`  that includes `After` this instance will not be used. See what I did there https://github.com/jaredpalmer/after.js/pull/173#issuecomment-443011616

Also it would be nice to support multiple components with `getInitialProps`. If you include a `<Footer/>` inside `<App/>` you might want to be able to load some data.

**Comment by jaredpalmer** [comment]
woop i meant to put this on #208

**Comment by jaredpalmer** [comment]
Lets move this conversation to #208 as it is further along.

**Pull request closed without merging**